### PR TITLE
Add hintable fields support

### PR DESCRIPTION
**Pull request by mvorisek**
see how to use in https://github.com/atk4/ui/pull/1559

then we can convert all tests to it

### CODE REVIEW

**Comment by mvorisek** [comment]
@georgehristov can you approve? I will then add to more tests and adjust the ui PR.

**Comment by DarkSide666** [comment]
What license is for `mahalux/atk4-hintable` ?

**Comment by mvorisek** [comment]
> 
> 
> What license is for `mahalux/atk4-hintable` ?

free for use as lib, donations welcomed if helpful for corporate clients :)

File: tests/ContainsManyTest.php [hunk 1]
```diff
@@ -51,8 +52,8 @@ protected function init(): void
         // discounts_total_sum - calculated by php callback not by SQL expression
         $this->addCalculatedField('discounts_total_sum', function (self $m) {
             $total = 0;
-            foreach ($m->ref('lines') as $line) {
-                $total += $line->get('total_gross') * $line->get('discounts_percent') / 100;
+            foreach ($m->lines as $line) {
+                $total += $line->total_gross * $line->get('discounts_percent') / 100;
```
Comment (DarkSide666) [inline tests/ContainsManyTest.php]: could be also `$line->discounts_percent`, right?

Comment (mvorisek) [reply tests/ContainsManyTest.php]: Of course! About 1000 more replaces ;-)

Working with PHPStan team to fix Model iterator support, then probably replace everything with Rector.

**Review by DarkSide666 — APPROVED** [review]
Real magic, but looks good to use and interesting as a bit different coding example.
Also important that it's 100% BC, so if you don't want to use this magic in your code - no worries.

**Comment by georgehristov** [comment]
Really good @mvorisek. 
My only comment would be that it is preferrable to have this into data directly to avoid dependency on external package.

**Comment by DarkSide666** [comment]
> Really good @mvorisek.
> My only comment would be that it is preferrable to have this into data directly to avoid dependency on external package.

Probably not in data directly, but in atk4/core because we can maybe use similar approach for atk4/ui controls too.

File: tests/ContainsManyTest.php (hunk 1 above)
Comment (mvorisek) [reply tests/ContainsManyTest.php]: discussion here: https://github.com/phpstan/phpstan/discussions/4220#discussioncomment-191882

**Pull request merged**